File: ai-mock/ai/point.py
# -*- coding: utf-8 -*-


class Point:
    def __init__(self, y, x):
        self.y = int(y)
        self.x = int(x)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __hash__(self):
        return self.x + 1000 * self.y

    def __add__(self, other):
        return Point(self.y + other.y, self.x + other.x)

    def __repr__(self):
        arrows = {
            Point(-1, 0): "^",
            Point(0, -1): "<",
            Point(0, 1): ">",
            Point(1, 0): "v",
            Point(0, 0): " ",
        }
        if self in arrows:
            return arrows[self]
        
        return "P({}, {})".format(self.y, self.x)
    
    def __lt__(self, other):
        if self.y == other.y:
            return self.x < other.x
        return self.y < other.y

    def dist(self, other):
        return abs(self.y - other.y) + abs(self.x - other.x)

File: ai/src/point.cpp
#include "point.h"

Point::Point() {
    y = -1;
    x = -1;
}

Point::Point(int _y, int _x) {
    y = _y;
    x = _x;
}

bool Point::operator==(const Point &other) const {
    return y == other.y && x == other.x;
}

bool Point::operator!=(const Point &other) const {
    return !(*this == other);
}

Point Point::operator+(const Point &other) const {
    return Point(y + other.y, x + other.x);
}

Point &Point::operator+=(const Point &other) {
    y += other.y;
    x += other.x;
    return *this;
}

Point Point::operator-(const Point &other) const {
    return Point(y - other.y, x - other.x);
}

bool Point::operator<(const Point &other) const {
    if (y == other.y) {
        return x < other.x;
    }
    return y < other.y;
}

int Point::dist(const Point &other) const {
    return abs(y - other.y) + abs(x - other.x);
}

string Point::print() const {
    if (abs(x) + abs(y) == 1) {
        if (x == 1) return "→";
        if (x == -1) return "←";
        if (y == 1) return "↓";
        if (y == -1) return "↑";
    }
    return "P(" + to_string(y) + ", " + to_string(x) + ")";
}

bool Point::isInsideField() const {
    return 0 <= y && y < Y && 0 <= x && x < X;
}